docs/design/graceful-domain-failover/3051-graceful-domain-failover.md

```diff
@@ -0,0 +1,239 @@
+# Design doc: Cadence graceful domain failover
+
+Author: Cadence Team
+
+Last updated: Mar 2020
+
+Reference: [#3051](https://github.com/uber/cadence/issues/3051)
+
+
+## Abstract
+
+Cadence supports domain failover with multi-cluster setup. However, the problems with the current failover are:
+1. Workflow progress could be lost.
+2. No causal consistency guarantee on domain level.
+
+The graceful domain failover uses to solve those two problems.
+
+## Use cases
+
+Users can trigger graceful domain failover via Cadence CLI (timeout is optional):
+
+`cadence --domain cadence-global domain update -failover_type graceful -active_cluster XYZ -failover_timeout 120s`
+
+Users can force complete domain failover via Cadence CLI:
+
+`cadence --domain cadence-global domain update -failover_type force -active_cluster XYZ`
+
+## Prerequisites
+There are conditions before starting a graceful domain failover.
+1. All clusters are available.
+2. The networking between clusters are available.
+3. The domain state has to be stable prior to graceful failover.
+
+## Limitation
+No concurrent graceful domain failover will be supported.
+
+## Proposal
+
+Due to the complexity of the protocol, it will go through the architecture from cross-cluster level, single cluster level to host level.
+
+The basic protocol is to insert markers in the active-to-passive cluster to indicate the boundary when the domain switches to passive. On the other side, the passive-to-active cluster listens to those markers and switches domain to active after receiving all the markers.
+
+### Cross cluster level
+
+![cross clusters sequence diagram](3051-cross-clusters.png)
+1, The operator issues a graceful failover to the passive-to-active cluster.
+
+2 - 3, The passive-to-active cluster gets domain data from all other clusters for two purposes: 1. Make sure the network and clusters are available prior to start the graceful failover. 2. Make sure there is no ongoing failover.
```

Comment:
If we want to be able to give the user an error message if they try to initiation concurrent graceful failovers then the before cluster X can respond back to cluster Y it needs to persist some data saying that it has responded yes to being able to handle a graceful failover. If it gets another attempt for graceful failover it needs to read this value and reject the request. Or we just say that users can initiate concurrent graceful failover but they lose all guarantees.

Reply:
Yes. We can do optimization to block concurrent graceful failover. However, if the concurrent graceful failover happens, we provide no guarantees.